Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A base ByteBuffer "Slicer", this base one just splits by a separator.
+public struct ByteBufferSliceSplitIterator: IteratorProtocol {
+    
+    var byteBuffer: ByteBuffer
+    var separator: UInt8
+    let start: Int
+    let length: Int
+    
+    private var currentIndex: Int
+    
+    public typealias Element = ByteBuffer
+    mutating public func next() -> ByteBuffer? {
+        
+        return byteBuffer.withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            
+            let startPoint   = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            
+            var address      = startPoint.advanced(by: currentIndex)
+            let finalAddress =    address.advanced(by: length)
+            
+            let initialPointer = address
+
+            while address < finalAddress && address.pointee != 0 {
+                
+                if address.pointee == separator {
+                    break
+                }
+                address = address.advanced(by: 1)
+            }
+            
+            let tokenLength = address - initialPointer
+            let ret = byteBuffer.getSlice(at: currentIndex, length: tokenLength)
+            // For the next time, with skipping the separator
+            currentIndex += tokenLength + 1
+            
+            return ret
+        }
+
+    }
+    
+    public init(byteBuffer: ByteBuffer, separator: UInt8, start: Int, length: Int) {
+        // This should never happens as we control when this is called. Adding an assert to ensure this.
+        assert(start <= byteBuffer.capacity - length)
+        
+        self.byteBuffer = byteBuffer
+        self.separator = separator
+        self.start = start
+        self.length = length
+        self.currentIndex = self.start
+    }
+
+    public init(byteBuffer: ByteBuffer,
+                separator: UInt8) {
+        
+        self.init(byteBuffer: byteBuffer, separator: separator,
+                  start: byteBuffer.readerIndex, length: byteBuffer.readableBytes)
+    }
+}
+
+extension ByteBuffer {
+    
+    public func sliceByTrimming(whiteSpaces: [UInt8], start: Int, length: Int) -> ByteBuffer? {
+        return withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            assert(start <= self.capacity - length)
```

Comment:
As this is a public method, this needs to be a precondition. It can also be outside of the closure.